src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Cohost/DocumentSnapshotFactory.cs

```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Composition;
+using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Cohost;
+
+[Export(typeof(DocumentSnapshotFactory)), Shared]
+[method: ImportingConstructor]
+internal class DocumentSnapshotFactory(Lazy<ProjectSnapshotFactory> projectSnapshotFactory)
+{
+    private readonly Lazy<ProjectSnapshotFactory> _projectSnapshotFactory = projectSnapshotFactory;
+
+    private readonly ConditionalWeakTable<TextDocument, IDocumentSnapshot> _documentSnapshots = new();
```

Comment:
I have never used a ConditionalWeakTable before, so if there is a better approach here I'm all ears. I will say that this should be temporary, because as we move further towards cohosting being the only game in town, I would expect `IDocumentSnaphot` to get a lot smaller, and possibly disappear altogether, so we might be able to just end up with a method for getting the generated context/host outputs from an extension method on the `TextDocument` itself, and let Roslyn do all of this state tracking.

Reply:
❗I recommend making this a static field. You shouldn't run into problems with this CWT in particular, but there can be a leak if any key in the CWT transitively references the CWT itself. I've seen that happen a few times over the years, and it can be difficult to spot. However, the problem can be easily averted by just making the CWT static.

Reply:
Thank you, will do